Generalise grid, gridLines, and tableLines to produce StringBuilders.

Xitian9 opened the pull request:
Instead of Strings.

`gridString` and `tableString` have been left producing Strings because they have ‘String’ in their names, however it would also be reasonable to generalise these.

Xitian9 committed:
Generalise grid, gridLines, and tableLines to produce StringBuilders
instead of Strings

muesli4 commented:
The versions ending in `String` were more a conceptual string but not fixed to the concrete type.  My initial concern for not turning this yet into `StringBuilder` was that I did not want users to force manually specializing.  But perhaps writing `:: String` is acceptable.  Thus, I would ask you to also convert those.

I'm willing to break the API for version `1.0.0.0`.  Perhaps we can get some other changes in to make it worth the effort. :)

Xitian9 commented:
I'm happy to do that if you want.

In hledger we've used the convention of having a main builder whose name ends with `B`, and a `String`-specialised version without the `B`. I don't think it would be a big issue to keep the legacy name for the specialised version, and have `gridBuilder` and `tableBuilder` for the generalised versions, if you wanted that.

Xitian9 commented:
I've made the changes you requested. Let me know if you'd like any further.

muesli4 commented:
What I would probably do is provide a special version of every function with a `B` suffix.  I like the idea.  In general, I don't know which of the functions are actually used.  But it is not too much effort for me to maintain and keep them, so I left it the way it is.

Xitian9 commented:
> What I would probably do is provide a special version of every function with a `B` suffix. I like the idea. In general, I don't know which of the functions are actually used. But it is not too much effort for me to maintain and keep them, so I left it the way it is.

Which functions do you mean by ‘every function’? Do you just mean `gridString` and `tableString`, or do you also mean `grid`, `gridLines`, and `tableLines`?

Xitian9 committed:
Add versions of table and grid functions specialised to Strings

The following functions now come in pairs: one which produces
StringBuilders, and another which produces Strings.

gridB <-> grid
gridLinesB <-> gridLines
gridStringB <-> gridString
tableLinesB <-> tableLines
tableStringB <-> tableString

muesli4 commented:
Please use imperative in commit messages.  Otherwise, seems fine to me, thanks a lot.

Xitian9 commented:
Commit message updated.

muesli4 commented:
I resolved the conflicts locally and rebased manually.
